Lib/test/test_capi/test_opt.py

```diff
@@ -1646,6 +1646,27 @@ def f(n):
         self.assertIn("_TO_BOOL_STR", uops)
         self.assertNotIn("_GUARD_TOS_UNICODE", uops)
 
+    def test_binary_subcsr_str_int_narrows_to_str(self):
+        def testfunc(n):
+            x = 0
+            s = "foo"
+            z = ""
+            for _ in range(n):
+                y = s[0]       # _BINARY_OP_SUBSCR_STR_INT
+                z = "bar" + y  # (_GUARD_TOS_UNICODE) + _BINARY_OP_ADD_UNICODE
+                x += 1
+            return x
+
+        res, ex = self._run_with_optimizer(testfunc, TIER2_THRESHOLD)
+        self.assertEqual(res, TIER2_THRESHOLD)
```

Comment:
Just a little bit of cleanup here. I'd prefer if the return value helped indicate that the code is still behaving as intended, and appending the result each time is a decent way of doing this. Also, we don't need the initialization of `z`:

```suggestion
        def testfunc(n):
            x = []
            s = "foo"
            for _ in range(n):
                y = s[0]       # _BINARY_OP_SUBSCR_STR_INT
                z = "bar" + y  # (_GUARD_TOS_UNICODE) + _BINARY_OP_ADD_UNICODE
                x.append(z)
            return x

        res, ex = self._run_with_optimizer(testfunc, TIER2_THRESHOLD)
        self.assertEqual(res, ["barf"] * TIER2_THRESHOLD)
```

Reply:
Nice, this suggestion looks much better than what I wrote! I think now it's clearer to me how the tests should be structured :)